### PR TITLE
 8268405: Several regressions 4-17% after CHA changes

### PR DESCRIPTION
**Pull request by iwanowww**
Recevier validation code in `Compile::optimize_inlining()` added by JDK-8267807 is way too strict. In particular, it rejects perfectly valid cases when unique concrete method comes from a superclass ( `receiver_type->klass() <: 
 cha_monomorphic_target->holder()`).  

Instead of reworking the validation logic, I decided to completely get rid of it and follow John's @rose00 suggestion [1] to perform receiver checks at runtime when needed. As I described earlier [2], receiver checks should never fail unless there's a product bug lurking in the code. Debug builds will assert if it ever happens (controlled by `-XX:+-VerifyReceiverTypes`), but product builds will deoptimize at the call site instead.

(The fix depends on #55 to avoid assertion failures in `Parse::do_field_access()`.)

Testing:
- [x] hs-tier1 - hs-tier6
- [x] manual performance runs (dacapo-jython, in particular)  

[1] https://github.com/openjdk/jdk/pull/4212#discussion_r643467030
[2] https://github.com/openjdk/jdk/pull/4212#discussion_r644377036

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8268405](https://bugs.openjdk.java.net/browse/JDK-8268405): Several regressions 4-17% after CHA changes

### Reviewers
 * [Vladimir Kozlov](https://openjdk.java.net/census#kvn) (@vnkozlov - **Reviewer**)
 * [Dean Long](https://openjdk.java.net/census#dlong) (@dean-long - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk17 pull/60/head:pull/60` \
`$ git checkout pull/60`

Update a local copy of the PR: \
`$ git checkout pull/60` \
`$ git pull https://git.openjdk.java.net/jdk17 pull/60/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 60`

View PR using the GUI difftool: \
`$ git pr show -t 60`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk17/pull/60.diff">https://git.openjdk.java.net/jdk17/pull/60.diff</a>

</details>

### CODE REVIEW

File: src/hotspot/share/opto/doCall.cpp [hunk 1]
```diff
@@ -164,6 +167,12 @@ CallGenerator* Compile::call_generator(ciMethod* callee, int vtable_index, bool
       bool should_delay = false;
       if (ilt->ok_to_inline(callee, jvms, profile, should_delay)) {
         CallGenerator* cg = CallGenerator::for_inline(callee, expected_uses);
+        if (cg != NULL && is_virtual_or_interface && !callee->is_static()) {
+          CallGenerator* trap_cg = CallGenerator::for_uncommon_trap(callee,
```
Comment (vnkozlov) [inline src/hotspot/share/opto/doCall.cpp:171]: Add comment here and in next changed code about why you want to deoptimize.

**Review by vnkozlov — APPROVED** [review]
Just one comment.

**Review by vnkozlov — APPROVED** [review]
Good.

**Comment by dean-long** [comment]
Did you run the Renaissance benchmarks that showed a regression?

**Review by dean-long — APPROVED** [review]

**Comment by iwanowww** [comment]
> Did you run the Renaissance benchmarks that showed a regression?

Yes, the fix restores the performance back on affected Renaissance benchmarks.

**Comment by iwanowww** [comment]
Thanks for the reviews, Vladimir and Dean.

/integrate

**Pull request closed without merging**